Override template source maps

zarend opened the pull request:
FYI: I'm putting this PR on the back-burner for now until we land [40585](https://github.com/angular/angular/pull/40585). That's because @alxhub deleted `overrideComponentTemplate` 💥 , which means much of this PR will need to be rewritten. Ty all for reviewing. I'll re-request reviews when this is ready again.

This is a 2-part PR about template source maps. The primary goal here is to cleanup our code for template sourcemaps. The secondary goal is to fix our diagnostics created when overriding a template. Commit 1 is a big commit that does a refactor. Commit 2 is the smaller commit which fixes the diagnostics override template produces and validates that the refactor is worthwhile.

## fix(language-service): don't hardcode 'override.html' for override template diagnostics

This fixes how we create diagnostics in the override template flow. It


does not affect template diagnostics created any other way.

This is the previous behavior when a template is overriden with a parse
error:
'Parser Error: Bindings cannot contain assignments at column 6 in [{{nope=false}}] in override.html@0:0'

This is the corrected error.
'Parser Error: Bindings cannot contain assignments at column 6 in [{{nope=false}}] in /app.html@0:0'

petebacondarwin commented on packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import {ConstantPool} from '@angular/compiler';
+import {TemplateSourceRegistryImpl} from '@angular/compiler-cli/src/ngtsc/annotations/src/template_source_registry';
```
drive-by comment: imports from ngcc to ngtsc should be relative and target the barrel for that area

```suggestion
import {TemplateSourceRegistryImpl} from '../../../src/ngtsc/annotations';
```

zarend replied:
thanks, I fixed the imports

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:
```diff
@@ -0,0 +1,120 @@
+ * Computes the line start map of the given text. This can be used in order to
+ * retrieve the line and character of a given text position index.
+ */
+export function computeLineStartsMap(text: string): number[] {
```
There is actually a function in ngtsc/sourcemaps that does this. Perhaps we should export that from the `sourcemaps` barrel and use it here instead of redefining it?

See https://github.com/angular/angular/blob/3cfb4b4ef9127c0eca34ad2edf93004a0eaab4c5/packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file.ts#L421

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:
```diff
@@ -0,0 +1,120 @@
+function findClosestLineStartPosition<T>(
+    linesMap: T[], position: T, low = 0, high = linesMap.length - 1) {
+  while (low <= high) {
+    const pivotIdx = Math.floor((low + high) / 2);
```
```suggestion
    const pivotIdx = (low + high) >> 1;
```

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:
```diff
@@ -0,0 +1,120 @@
+ * Represents the source of a template that was processed during type-checking. This information is
+ * used when translating parse offsets in diagnostics back to their original line/column location.
+ */
+// TODO: delete me, this is duplicated from template_source_registry.ts
```
This *is* the template_source_registry.ts 😸 !

zarend replied:
`// TODO: delete irrelevant comments` 😂

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/template_source_registry.ts:
```diff
@@ -0,0 +1,120 @@
+
+  captureSource(node: ClassDeclaration, mapping: TemplateSourceMapping, file: ParseSourceFile):
+      ClassDeclaration {
+    this.templateSources.set(node, new TemplateSource(mapping, file));
```
Does this mean that a new `TemplateSource` will be generated for every class declaration in a file? Perhaps this should be cached here based on the `file`?

zarend commented:
### TODO
There are the items todo before this PR will be ready to review.
 - sort out `ts.ClassDeclaration` vs the `ClassDeclaration` from reflection
 - fix TODO's in code comments
 - add unit tests for inline and indirect templates
 - squash into a single commit

zarend commented:
TODO move TemplateSourceRegistry to typecheck directory

zarend committed:
refactor(compiler-cli): add api to lookup template source maps by class declaration

This is about refactoring the different ways to find the source mapping
for a template. TemplateSourceManager allows us to lookup source maps by
template id, but does not allow us to lookup by class declaration.

This renames TemplateSourceManager to TemplateNodeManager and changes it
to map template id to template declaration. It adds a new class called
TemplateSourceRegistry which maps class declaration to source map.

The trade-off is that we are now able to lookup source maps by class
declaration, but it takes 2 steps to lookup by template id. First we
lookup the class declaration by tempalte id via TemplateNodeManager.
Then we lookup the sourcemap in TemplateSourceRegistry.

zarend committed:
fix(language-service): don't hardcode 'override.html' for override
template diagnostics

This fixes how we create diagnostics in the override template flow. It
does not affect template diagnostics created any other way.

This is the previous behavior when a template is overriden with a parse
error:
'Parser Error: Bindings cannot contain assignments at column 6 in [{{nope=false}}] in override.html@0:0'

This is the corrected error.
'Parser Error: Bindings cannot contain assignments at column 6 in [{{nope=false}}] in /app.html@0:0'

petebacondarwin commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -154,7 +154,33 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
   overrideComponentTemplate(component: ts.ClassDeclaration, template: string):
       {nodes: TmplAstNode[], errors: ParseError[]|null} {
-    const {nodes, errors} = parseTemplate(template, 'override.html', {
+    const sourceMapping = this.templateSourceRegistry.getSourceMapping(component);
+    let templateUrl: string;
+
+    if (sourceMapping === null) {
+      // A source mapping could be unavailable for a broken component. Fallback to 'override.html'
+      // as an error state.
+      templateUrl = 'override.html';
+    } else {
+      switch (sourceMapping.type) {
+        case 'direct': {
+          templateUrl = absoluteFromSourceFile(component.getSourceFile());
+          break;
+        }
+        case 'indirect': {
+          // Fallback to 'override.html' because we cannot correctly sourcemap an indirect
+          // template.
+          templateUrl = 'override.html';
+          break;
+        }
+        case 'external': {
+          templateUrl = sourceMapping.templateUrl;
+          break;
+        }
+      }
+    }
```
NIT: call be old fashioned but I would be keen to put this whole block into its own helper function:

```ts
const templateUrl = getTemplateUrl(component);
```

petebacondarwin commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -209,7 +210,8 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
     return getTemplateMapping(
-        shimSf, positionInShimFile, fileRecord.sourceManager, /*isDiagnosticsRequest*/ false);
+        shimSf, positionInShimFile, fileRecord.sourceManager, /*isDiagnosticsRequest*/ false,
+        this.getTemplateSourceRegistry());
```
OOC why use the `getTemplatSourceRegistry()` in these functions rather than just refer to the private `templateSourceRegistry` directly?

petebacondarwin commented on packages/compiler-cli/src/ngtsc/typecheck/src/template_source_registry.ts:
```diff
@@ -0,0 +1,143 @@
+
+  captureSource(node: ClassDeclaration, mapping: TemplateSourceMapping, file: ParseSourceFile):
+      ClassDeclaration {
+    this.templateSources.set(node, new TemplateSource(mapping, file));
```
This will still compute and store the line-starts of the `file` for every `node` captured from the file. Perhaps the line-starts should be stored on the `file` itself, to avoid having to do this computation redundantly?

petebacondarwin reviewed (commented):
Nice refactoring @zarend !

atscott reviewed (changes requested):
Just some minor nits. Nice work!

zarend commented:
FYI: I'm putting this PR on the back-burner for now until we land [40585](https://github.com/angular/angular/pull/40585). That's because @alxhub deleted `overrideComponentTemplate` 💥 , which means much of this PR will need to be rewritten. Ty all for reviewing. I'll re-request reviews when this is ready again.

petebacondarwin commented:
@zarend - #40585 has now landed. Is it time to restart this PR?
